pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,250 @@
+	ErrorLogMaxAge      int
+	ErrorLogFileMaxSize int
+
+	errorChannel chan zapcore.Entry
+	queue        []zapcore.Entry
+	Ticker       *time.Ticker
+	errorFile    *os.File
+	size         int
+}
+
+// NewECSErrorLogger could new a struct based on the 3 environment variables:
+// error file path, error log TTL and error file max size
+func NewECSErrorLogger() *ECSErrorLogger {
+	ecsErrorLogFilePath := os.Getenv("STATUS_MESSAGE_FILE_PATH")
+	errorLogMaxAge, err := strconv.Atoi(os.Getenv("STATUS_MESSAGE_TTL"))
+	if err != nil || errorLogMaxAge <= 0 {
+		log.Fatalf("failed to get STATUS_MESSAGE_TTL: %v", err)
+	}
+	errorFileSize, err := strconv.Atoi(os.Getenv("STATUS_MESSAGE_MAX_BYTE_LENGTH"))
+	if err != nil || errorFileSize <= 0 {
+		log.Fatalf("failed to get STATUS_MESSAGE_MAX_BYTE_LENGTH: %v", err)
+	}
+	if errorFileSize > 1000 {
+		log.Printf("error log file size should not exceed 1KB, will use 1000 instead")
+		errorFileSize = 1000
+	}
+	return &ECSErrorLogger{
+		ErrorLogFilePath:    ecsErrorLogFilePath,
+		ErrorLogMaxAge:      errorLogMaxAge,
+		ErrorLogFileMaxSize: errorFileSize,
+		errorChannel:        make(chan zapcore.Entry, 100),
+	}
+}
+
+// Run func help select the different case between either write error log into file or
+// rotate the old logs in file based on TTL
+func (l *ECSErrorLogger) Run() {
+	defer l.Ticker.Stop()
+	for {
+		select {
+		case newError := <-l.errorChannel:
+			l.Write(newError)
+		case <-l.Ticker.C:
+			l.processTimeout()
+		}
+	}
+}
+
+// Write function could write the new error into the file, when the file size plus new error size
+// exceed the max limit, will call rotate function to delete the old log
+func (l *ECSErrorLogger) Write(newError zapcore.Entry) {
+	if l.errorFile == nil {
+		if err := l.openExistingOrNewErrorFile(); err != nil {
+			log.Printf("[ecs error reporter] could not open error log file when write new error, err: %v", err)
+			return
+		}
+	}
+	newErrorLog := []byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+		newError.Time, newError.Level, newError.Caller, newError.Message))
+
+	if len(newErrorLog) > l.ErrorLogFileMaxSize {
+		log.Printf("[ecs error reporter] error size exceed the max size of error file")
+		return
+	}
+
+	err := l.errorFile.Truncate(0)
+	if err != nil {
+		log.Printf("[ecs error reporter] could not truncate error log file when write new error, err: %v", err)
+		return
+	}
+	l.size = 0
+
+	errorNum := len(l.queue)
+	for i := 0; i < errorNum; i++ {
+		e := l.queue[0]
+		if e.Message != newError.Message {
+			n, err := l.errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				log.Printf("[ecs error reporter] could not write error into error log file, err: %v", err)
+				return
+			}
+			l.size += n
+			l.queue = append(l.queue, e)
+		}
+		l.queue = l.queue[1:]
+	}
+
+	if l.size+len(newErrorLog) > l.ErrorLogFileMaxSize {
+		l.rotate(len(newErrorLog))
+	}
```

Comment:
Can't we know this before we attempt to rewrite the log file above?  This can result in truncating and rewriting the log file twice per error message with messages as small as 512 bytes (smaller in practice since the formatting around the message is ~85 bytes at minimum).